lib/mix/lib/mix.ex

```diff
@@ -675,47 +699,75 @@ defmodule Mix do
         Application.put_all_env(config, persistent: true)
         System.put_env(system_env)
 
-        installs_root =
-          System.get_env("MIX_INSTALL_DIR") || Path.join(Mix.Utils.mix_cache(), "installs")
```

Comment:
`MIX_INSTALL_DIR` is public. We cannot remove this.

The only thing you can't determinate is the `id`, so I suggest adding a `:cache_id` option?